File: plugin/zeppelin/src/main/scala/org/intellij/plugin/zeppelin/api/websocket/WebSocketModels.scala
package org.intellij.plugin.zeppelin.api.websocket

import org.intellij.plugin.zeppelin.models.Credentials
import spray.json.{DefaultJsonProtocol, DeserializationException, JsObject, JsString, JsValue, RootJsonFormat}


case class RunParagraphData(id: String,
                            paragraph: Option[String],
                            title: Option[String],
                            settings: Map[String, String] = Map(),
                            params: Map[String, String] = Map(),
                            config: Map[String, String] = Map())

object ZeppelinWebSocketProtocol extends DefaultJsonProtocol {
  implicit val RunParagraphFormat: RootJsonFormat[RunParagraphData] = jsonFormat6(RunParagraphData)
}


/**
  * The response web socket codes
  */
object ResponseCode extends Enumeration {
  type ResponseCode = Value
  val PARAGRAPH_UPDATE_OUTPUT, PARAGRAPH_APPEND_OUTPUT, PARAGRAPH, PARAGRAPH_ADDED, NOTE, PROGRESS,
  SAVE_NOTE_FORMS, NOTES_INFO, INTERPRETER_BINDINGS, PARAS_INFO = Value
}

case class SessionIsClosedException() extends Exception()

case class RequestMessage(op: String, data: JsValue, credentials: Credentials)

case class ResponseMessage(op: String, data: JsObject)

trait MessageHandler {
  def handle(result: ResponseMessage)
}

object WebSocketApiProtocol extends DefaultJsonProtocol {

  implicit object RequestMessageFormat extends RootJsonFormat[RequestMessage] {
    def read(value: JsValue): RequestMessage = {
      throw DeserializationException("Non implemented")
    }

    def write(r: RequestMessage): JsObject = {
      JsObject(
        "op" -> JsString(r.op),
        "data" -> r.data,
        "ticket" -> JsString(r.credentials.ticket),
        "principal" -> JsString(r.credentials.principal),
        "roles" -> JsString(r.credentials.roles)
      )
    }
  }


  implicit object ResponseMessageFormat extends RootJsonFormat[ResponseMessage] {
    def read(value: JsValue): ResponseMessage = {
      value.asJsObject.getFields("op", "data") match {
        case Seq(JsString(op), data) => ResponseMessage(op, data.asJsObject())
        case _ => throw DeserializationException("Response message expected")
      }
    }

    def write(r: ResponseMessage): JsValue = {
      throw throw DeserializationException("Non implemented")
    }
  }

}

File: zeppelin/src/org/intellij/plugin/zeppelin/api/websocket/WebSocketModels.kt
package org.intellij.plugin.zeppelin.api.websocket

import org.intellij.plugin.zeppelin.models.Credentials

data class RunParagraphData(val id: String,
                            val paragraph: String,
                            val title: String,
                            val settings: Map<String, String> = emptyMap(),
                            val params: Map<String, String> = emptyMap(),
                            val config: Map<String, String> = emptyMap())

/**
 * The zeppelin response web socket codes
 */
enum class ResponseCode {
    PARAGRAPH_UPDATE_OUTPUT,
    PARAGRAPH_APPEND_OUTPUT,
    PARAGRAPH,
    PARAGRAPH_ADDED,
    NOTE,
    PROGRESS,
    SAVE_NOTE_FORMS,
    NOTES_INFO,
    INTERPRETER_BINDINGS,
    PARAS_INFO
}

data class RequestMessage(val op: String, val data: String, val credentials: Credentials)

data class ResponseMessage(val op: String, val data: String)

interface MessageHandler {
    fun handle(result: ResponseMessage)
}

/**
 * The response web socket codes
 */
enum class RequestOperations {
    GET_NOTE, RUN_PARAGRAPH, GET_INTERPRETER_BINDINGS, SAVE_INTERPRETER_BINDINGS
}

